### PR TITLE
Add the old tag name to tag_renamed hook

### PR DESCRIPTION
**Pull request by blarz**
As stated correctly in #999, this hook would be more useful
if you'd know which tag was renamed.

### CODE REVIEW

**Comment by blarz** [comment]
Adding a static variable to the `tag->name.changed()` lambda was the only way I could see, to carry the old name with me somehow. Maybe somebody else has a better idea?

The `Attribute_::changed` signal doesn't give us the old value, so I had to figure out another way to keep track of the old value.

**Comment by blarz** [comment]
Also the `replace()` dance I do in the test feels a bit weird, maybe there is a more elegant solution there, too.

**Comment by t-wissmann** [comment]
Thanks for fixing these open bugs :)

> Adding a static variable to the `tag->name.changed()` lambda was the only way I could see, to carry the old name with me somehow. Maybe somebody else has a better idea?

The `static` creates only one variable for this lambda (and not for one closure!) and so the `static` leads to wrong results when there is more than one tag. I've pushed a test case where I do the renaming:
```
foo_old -> foo_new
bar_old -> bar_new
```
but the second call creates the hook `tag_renamed foo_new bar_new`, because the static variable is set to `foo_new`.

I think the best way is to have a private member in the `Tag` class `oldName_` that is used here. By the way, in the "Validator"-callbacks, the value of the attribute is still the old one, but I think it's cleaner to run this hook emision in the `changed()` callback.

Another thing regarding compatibility: This PR changes the semantics from `tag_renamed new` to `tag_renamed old new`, so if someone has code that listens for it and uses the first parameter, then the code now gets the old instead of the new tag name. While I think it is fine to introduce this incompatibility since the order "old new" makes much more sense, we should mention this in the `MIGRATION` file.

**Comment by t-wissmann** [comment]
> Also the `replace()` dance I do in the test feels a bit weird, maybe there is a more elegant solution there, too.

I've made the entries in the `RENAME_COMMANDS` array lambda functions, so one can call
```python
hc.call(rename_command('old', 'new'))
```
If you don't like it feel free to revert my commit 44c6306 (and it thus will disappear in the squashed commit for this PR)

File: doc/herbstluftwm.txt [hunk 1]
```diff
@@ -1667,6 +1667,9 @@ tag_added 'TAG'::
 tag_removed 'TAG'::
     The tag named 'TAG' was removed.
 
+tag_renamed 'OLD' 'NEW::
```
Comment (tiosgz) [inline doc/herbstluftwm.txt:1670]: Missing `'`?

File: tests/test_tags.py [hunk 2]
```diff
@@ -256,3 +261,17 @@ def test_urgent_count(hlwm, x11):
 
     # since one of them gets focused, 4 urgent clients remain
     assert int(hlwm.get_attr('tags.focus.urgent_count')) == 4
+
+
+def test_rename_different_tags(hlwm, hc_idle):
```
Comment (tiosgz) [inline tests/test_tags.py:266]: I think `test_rename_multiple_tags` is a better name.

File: doc/herbstluftwm.txt (hunk 1 above)
Comment (blarz) [reply doc/herbstluftwm.txt]: Thanks!

File: tests/test_tags.py (hunk 2 above)
Comment (blarz) [reply tests/test_tags.py]: True, changed that. Thanks!

**Comment by blarz** [comment]
> If you don't like it feel free to revert my commit [44c6306](https://github.com/herbstluftwm/herbstluftwm/commit/44c6306569b3dfeaeca38e11313e5dbca59c73ef) (and it thus will disappear in the squashed commit for this PR)

Perfect, thank you!

**Comment by blarz** [comment]
> Thanks for fixing these open bugs :)

Sure thing! The autumn air outside inspired me to do some hlwm hacking :smile: 

> > Adding a static variable to the `tag->name.changed()` lambda was the only way I could see, to carry the old name with me somehow. Maybe somebody else has a better idea?
> 
> The `static` creates only one variable for this lambda (and not for one closure!) and so the `static` leads to wrong results when there is more than one tag. I've pushed a test case where I do the renaming:
> 
> ```
> foo_old -> foo_new
> bar_old -> bar_new
> ```
> 
> but the second call creates the hook `tag_renamed foo_new bar_new`, because the static variable is set to `foo_new`.

I knew there was something I've overlooked, thank you for the new test!

> I think the best way is to have a private member in the `Tag` class `oldName_` that is used here. By the way, in the "Validator"-callbacks, the value of the attribute is still the old one, but I think it's cleaner to run this hook emision in the `changed()` callback.

I'll try that, thanks!

> Another thing regarding compatibility: This PR changes the semantics from `tag_renamed new` to `tag_renamed old new`, so if someone has code that listens for it and uses the first parameter, then the code now gets the old instead of the new tag name. While I think it is fine to introduce this incompatibility since the order "old new" makes much more sense, we should mention this in the `MIGRATION` file.

I'll add an entry in the next version of this PR. Although as pointed out in #999 tag_renamed isn't even officially documented?

**Comment by t-wissmann** [comment]
> > I think the best way is to have a private member in the `Tag` class `oldName_` that is used here. By the way, in the "Validator"-callbacks, the value of the attribute is still the old one, but I think it's cleaner to run this hook emision in the `changed()` callback.
> 
> I'll try that, thanks!

As you noticed already, `private` doesn't work here; but I think `public` or `friend` are both OK.


> > Another thing regarding compatibility: This PR changes the semantics from `tag_renamed new` to `tag_renamed old new`, so if someone has code that listens for it and uses the first parameter, then the code now gets the old instead of the new tag name. While I think it is fine to introduce this incompatibility since the order "old new" makes much more sense, we should mention this in the `MIGRATION` file.
> 
> I'll add an entry in the next version of this PR. Although as pointed out in #999 tag_renamed isn't even officially documented?

I'd say we should put a warning there, since `tag_renamed` exists for quite a while already

**Comment by blarz** [comment]
> > > I think the best way is to have a private member in the `Tag` class `oldName_` that is used here. By the way, in the "Validator"-callbacks, the value of the attribute is still the old one, but I think it's cleaner to run this hook emision in the `changed()` callback.
> > 
> > 
> > I'll try that, thanks!
> 
> As you noticed already, `private` doesn't work here; but I think `public` or `friend` are both OK.

Yes, I struggled a bit but then added a public member :)

> > > Another thing regarding compatibility: This PR changes the semantics from `tag_renamed new` to `tag_renamed old new`, so if someone has code that listens for it and uses the first parameter, then the code now gets the old instead of the new tag name. While I think it is fine to introduce this incompatibility since the order "old new" makes much more sense, we should mention this in the `MIGRATION` file.
> > 
> > 
> > I'll add an entry in the next version of this PR. Although as pointed out in #999 tag_renamed isn't even officially documented?
> 
> I'd say we should put a warning there, since `tag_renamed` exists for quite a while already

Makes sense. I added an entry, although I put it under "current git version" because I don't know what the next version is (and also I don't know since when tag_renamed has been around with one parameter).

**Review by t-wissmann — APPROVED** [review]
Looks good, thanks!

**Pull request merged**